pkg/apis/kops/validation/validation.go

```diff
@@ -599,9 +603,75 @@ func validateNetworkingCalico(v *kops.CalicoNetworkingSpec, e *kops.EtcdClusterS
 		allErrs = append(allErrs, IsValidValue(fldPath.Child("iptablesBackend"), &v.IptablesBackend, valid)...)
 	}
 
+	if v.IPv4AutoDetectionMethod != "" {
+		allErrs = append(allErrs, validateCalicoAutoDetectionMethod(fldPath.Child("ipv4AutoDetectionMethod"), &v.IPv4AutoDetectionMethod, ipv4.Version)...)
+	}
+
+	if v.IPv6AutoDetectionMethod != "" {
+		allErrs = append(allErrs, validateCalicoAutoDetectionMethod(fldPath.Child("ipv6AutoDetectionMethod"), &v.IPv6AutoDetectionMethod, ipv6.Version)...)
+	}
+
 	return allErrs
 }
 
+func validateCalicoAutoDetectionMethod(fldPath *field.Path, runtime *string, version int) field.ErrorList {
+	methodFirstFound := "first-found"
+	methodCanReach := "can-reach="
+	methodInterface := "interface="
+	methodSkipInterface := "skip-interface="
+	validationError := field.ErrorList{}
+
+	// validation code is based on the checks in calico/node startup code
+	// valid formats are "first-found", "can-reach=DEST", or
+	// "(skip-)interface=<COMMA-SEPARATED LIST OF INTERFACES>"
+	//
+	// We won't do deep validation of the values in this check, since they can
+	// be actual interface names or regexes
+	if *runtime == methodFirstFound {
+		return field.ErrorList{}
+
+	} else if strings.HasPrefix(*runtime, methodCanReach) {
+		destStr := strings.TrimPrefix(*runtime, methodCanReach)
+		if version == ipv4.Version {
+			return utilvalidation.IsValidIPv4Address(fldPath, destStr)
+		} else if version == ipv6.Version {
+			return utilvalidation.IsValidIPv6Address(fldPath, destStr)
+		}
+
+		return field.ErrorList{field.Invalid(fldPath, runtime, "Invalid; IP version in incorrect")}
+
+	} else if strings.HasPrefix(*runtime, methodInterface) {
+		ifStr := strings.TrimPrefix(*runtime, methodInterface)
+		ifRegexes := regexp.MustCompile(`\s*,\s*`).Split(ifStr, -1)
+		if len(ifRegexes) == 0 || ifRegexes[0] == "" {
+			validationError = append(validationError, field.Invalid(fldPath, runtime, "Expected 'interface=<COMMA-SEPARATED-LIST>'"))
+		}
+		for _, r := range ifRegexes {
+			_, e := regexp.Compile(r)
+			if e != nil {
+				validationError = append(validationError, field.Invalid(fldPath, runtime, "Invalid regexp: "+e.Error()))
```

Comment:
Don't capitalize error strings. Avoid saying "Invalid" because that word will be earlier in the error message.

Reply:
Thanks for the better phrasing